File: src/test/java/spring/graphql/modules/product/ProductControllerTest.java
package spring.graphql.modules.product;

import org.junit.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.math.BigDecimal;
import java.util.List;
import java.util.UUID;

import static java.util.Optional.of;
import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.Mockito.*;

@ExtendWith(MockitoExtension.class)
public class ProductControllerTest {
    @Mock
    private ProductRepository productRepository;

    @Test
    public void products() {
        ProductEntity firstProduct = new ProductEntity();
        ProductEntity secondProduct = new ProductEntity();

        firstProduct.setName("First Product");
        firstProduct.setPrice(BigDecimal.ONE);
        firstProduct.setCategoryId(UUID.randomUUID());

        secondProduct.setName("Second Product");
        secondProduct.setPrice(BigDecimal.ONE);
        secondProduct.setCategoryId(UUID.randomUUID());

        when(productRepository.findAll()).thenReturn(List.of(firstProduct, secondProduct));

        Iterable<ProductEntity> result = productRepository.findAll();

        assertNotNull(result);
        result.forEach(productEntity -> {
            assertNotNull(productEntity);
            assertTrue(productEntity.getName().equals(firstProduct.getName()) || productEntity.getName().equals(secondProduct.getName()));
        });
    }

    @Test
    public void product() {
        ProductEntity productEntity = new ProductEntity();

        productEntity.setName("test");
        productEntity.setPrice(BigDecimal.ONE);
        productEntity.setCategoryId(UUID.randomUUID());

        when(productRepository.findById(productEntity.getId())).thenReturn(of(productEntity));

        ProductEntity result = productRepository.findById(productEntity.getId()).orElseThrow();

        assertNotNull(result);
        assertEquals(productEntity, result);
    }

    @Test
    public void createProduct() {
        ProductEntity productEntity = new ProductEntity();

        productEntity.setName("test");
        productEntity.setPrice(BigDecimal.ONE);
        productEntity.setCategoryId(UUID.randomUUID());

        when(productRepository.save(productEntity)).thenReturn(productEntity);

        ProductEntity result = productRepository.save(productEntity);

        assertNotNull(result);
        assertEquals(productEntity, result);
    }

    @Test
    public void updateProduct() {
        ProductEntity productEntity = new ProductEntity();

        productEntity.setName("test");
        productEntity.setPrice(BigDecimal.ONE);
        productEntity.setCategoryId(UUID.randomUUID());

        when(productRepository.save(productEntity)).thenReturn(productEntity);

        ProductEntity result = productRepository.save(productEntity);

        assertNotNull(result);
        assertEquals(productEntity, result);

        productEntity.setName("test2");
        productEntity.setPrice(BigDecimal.TEN);

        when(productRepository.save(productEntity)).thenReturn(productEntity);

        result = productRepository.save(productEntity);

        assertNotNull(result);
        assertEquals(productEntity, result);
    }

    @Test
    public void deleteProduct() {
        ProductEntity productEntity = new ProductEntity();

        productEntity.setName("test");
        productEntity.setPrice(BigDecimal.ONE);
        productEntity.setCategoryId(UUID.randomUUID());

        when(productRepository.save(productEntity)).thenReturn(productEntity);

        ProductEntity result = productRepository.save(productEntity);

        assertNotNull(result);
        assertEquals(productEntity, result);

        doNothing().when(productRepository).deleteById(productEntity.getId());
        productRepository.deleteById(productEntity.getId());

        verify(productRepository, times(1)).deleteById(productEntity.getId());
    }
}

File: src/test/kotlin/spring/graphql/modules/product/ProductControllerTest.kt
package spring.graphql.modules.product

import com.nextspring.modules.product.ProductEntity
import com.nextspring.modules.product.ProductRepository
import org.junit.Test
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.Mock
import org.mockito.Mockito
import org.mockito.junit.jupiter.MockitoExtension
import java.math.BigDecimal
import java.util.*
import java.util.function.Consumer

@ExtendWith(MockitoExtension::class)
class ProductControllerTest {
    @Mock
    private val productRepository: ProductRepository = Mockito.mock(ProductRepository::class.java)

    @Test
    fun products() {
        val firstProduct = ProductEntity()
        val secondProduct = ProductEntity()

        firstProduct.name = "First Product"
        firstProduct.price = BigDecimal.ONE

        secondProduct.name = "Second Product"
        secondProduct.price = BigDecimal.ONE

        Mockito.`when`(productRepository.findAll()).thenReturn(listOf(firstProduct, secondProduct))

        val result: Iterable<ProductEntity> = productRepository.findAll()

        Assertions.assertNotNull(result)
        result.forEach(Consumer { productEntity: ProductEntity ->
            Assertions.assertNotNull(productEntity)
            Assertions.assertTrue(productEntity.name == firstProduct.name || productEntity.name == secondProduct.name)
        })
    }

    @Test
    fun product() {
        val productEntity = ProductEntity()

        productEntity.id = UUID.randomUUID()
        productEntity.name = "test"
        productEntity.price = BigDecimal.ONE

        Mockito.`when`(productRepository.findById(productEntity.id!!)).thenReturn(Optional.of(productEntity))

        val result = productRepository.findById(productEntity.id!!).orElseThrow()

        Assertions.assertNotNull(result)
        Assertions.assertEquals(productEntity, result)
    }

    @Test
    fun createProduct() {
        val productEntity = ProductEntity()

        productEntity.name = "test"
        productEntity.price = BigDecimal.ONE

        Mockito.`when`(productRepository.save(productEntity)).thenReturn(productEntity)

        val result = productRepository.save(productEntity)

        Assertions.assertNotNull(result)
        Assertions.assertEquals(productEntity, result)
    }

    @Test
    fun updateProduct() {
        val productEntity = ProductEntity()

        productEntity.name = "test"
        productEntity.price = BigDecimal.ONE

        Mockito.`when`(productRepository.save(productEntity)).thenReturn(productEntity)

        var result = productRepository.save(productEntity)

        Assertions.assertNotNull(result)
        Assertions.assertEquals(productEntity, result)

        productEntity.name = "test2"
        productEntity.price = BigDecimal.TEN

        Mockito.`when`(productRepository.save(productEntity)).thenReturn(productEntity)

        result = productRepository.save(productEntity)

        Assertions.assertNotNull(result)
        Assertions.assertEquals(productEntity, result)
    }

    @Test
    fun deleteProduct() {
        val productEntity = ProductEntity()

        productEntity.id = UUID.randomUUID()
        productEntity.name = "test"
        productEntity.price = BigDecimal.ONE

        Mockito.`when`(productRepository.save(productEntity)).thenReturn(productEntity)

        val result = productRepository.save(productEntity)

        Assertions.assertNotNull(result)
        Assertions.assertEquals(productEntity, result)

        Mockito.doNothing().`when`(productRepository).deleteById(productEntity.id!!)
        productRepository.deleteById(productEntity.id!!)

        Mockito.verify(productRepository, Mockito.times(1)).deleteById(productEntity.id!!)
    }
}